ThunderCloud/TSCAppDelegate.swift

```diff
@@ -0,0 +1,196 @@
+			streamingController.fetchStreamingPage(cacheURLString: cacheURL, completion: { (stormViewController, error) in
+				
+				OperationQueue.main.addOperation {
+					
+					guard let window = self.window else { return }
+					
+					MDCHUDActivityView.finish(in: window)
+					
+					if let error = error {
+						UIAlertController.presentError(error, in: window.rootViewController!)
+					}
+					
+					guard let viewController = stormViewController else { return }
+					
+					viewController.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(self.dismissStreamedPage))
+					let navController = UINavigationController(rootViewController: viewController)
+					window.rootViewController?.present(navController, animated: true, completion: nil)
+				}
+			})
+			
+			return true
+			
+		} else {
+			
+			// Load the page locally
+			guard let pageURL = URL(string: cacheURL) else {
+				return false
+			}
+			
+			guard let viewController = TSCStormViewController(url: pageURL) else { return false }
+			
+			viewController.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: viewController, action: #selector(UIViewController.dismissAnimated))
+			let navController = UINavigationController(rootViewController: viewController)
+			window?.rootViewController?.present(navController, animated: true, completion: nil)
+			
+			return true
+		}
+	}
+	
+	public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+		
+		let application = UIApplication.shared
+		_ = handleNotificationResponse(for: response.notification, response: response, fromLaunch: application.applicationState == .inactive || application.applicationState == .background)
+		
+		completionHandler()
+	}
+	
+	open func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
+		
+		if let foregroundOptions = foregroundNotificationOptions {
+			completionHandler(foregroundOptions)
+			return
+		}
+		
+	    _ = handleNotificationResponse(for: notification, response: nil, fromLaunch: false)
+		completionHandler([])
+	}
+	
+	//MARK: -
+	//MARK: - Spotlight indexing
+	//MARK: -
+	
+	open func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
+		
+		guard let searchableItemIdentifier = userActivity.userInfo?[CSSearchableItemActivityIdentifier] as? String, searchableItemIdentifier.contains(".json") else {
+			return false
+		}
+		
+		guard let url = URL(string: "caches://pages/\(searchableItemIdentifier)") else { return false }
+		
+		guard let stormViewController = TSCStormViewController(url: url) else { return false }
```

Comment:
Could probably combine this guard with the one above, both return false and `url` from the first one is used here. Unsure whether it should all be rolled into the top-most guard, as then there'd be lots going on in just one guard and it'd be a bit unwieldy. I've often used a limit of 3 conditions per guard to keep it clean, what is your opinion on it? :)

Reply:
I've tried to avoid combining basically when it's going to make the line significantly longer than other lines around it. however I think in this case it should be fine 😀